### DISCUSSION TITLE
using FIR filter and volume control for a2dp_source

### DISCUSSION THREAD

**Discussion by mohaddeseh-rezaee** [pschatzmann/arduino-audio-tools#562] [Q&A] [answered]
Hi again, 
I was wondering if using features such as FIR filter and controlling volume is available for a2dp_source mode? 
Is it possible to control the volume with the BT headset? 
I am using this example and it works fine. [base-adc-a2dp](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/base-adc-a2dp/base-adc-a2dp.ino)
can I add [streams-audiokit-filter-audiokit.ino](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-filter-audiokit/streams-audiokit-filter-audiokit.ino) to previous code I mentioned ?
 I'm also wondering if it is possible to save data on sd card or no ? 
Thanks in advance!

**Comment by pschatzmann** [comment, marked as answer]
I don't see any issue here, so I convert it to a discussion

**Reply by pschatzmann** [reply]
To implement a stable A2DP source is quite challenging. So you should only consider to try to build this if you have plenty of experience!
Quite a few people had problems with the A2DPStream so I would recommend to use the basic API with the callbacks.

I recommend to handle the volume via the A2DP API.

And of cause you can build any type of audio processing chains doing any operation and writing data to any supported output which includes files.
